src/metabase/models/field_values.clj

```diff
@@ -234,14 +235,17 @@
   [field-or-field-id]
   (if-not (map? field-or-field-id)
     (let [field-id (u/the-id field-or-field-id)]
-      (recur (or (t2/select-one ['Field :base_type :visibility_type :has_field_values] :id field-id)
+      (recur (or (t2/select-one 'Field :id field-id)
                  (throw (ex-info (tru "Field {0} does not exist." field-id)
                                  {:field-id field-id, :status-code 404})))))
     (let [{base-type        :base_type
            visibility-type  :visibility_type
-           has-field-values :has_field_values} field-or-field-id]
+           has-field-values :has_field_values} field-or-field-id
+          table (t2/select-one 'Table :id (:table_id field-or-field-id))
```

Comment:
A few notes here

```suggestion
          table (t2/select-one-fn :db_id [:model/Table :db_id] :id (:table_id field-or-field-id))
```

1. Always use `(t2/select-one-fn <keyword> ...)` rather than `(<keyword> (t2/select-one ...))` because the first one only reads/transforms that specific column from the `ResultSet` whereas the latter has to read the entire row and create a map with every value which we then turn around and immediately discard. So its both significantly faster and more concise to use `select-one-fn` 
2. You should use `:model/Table` instead of `'Table` going forward, quoted model names are deprecated at this point
3. As another optimization you can do `[:model/Table :db_id]` instead of just `:model/Table`, that tells it to for real only fetch that one specific column from the app DB. Since `select-one-fn` only reads that column anyway it shouldn't make a huge difference but some JDBC drivers are dumb and fetch the entire row anyway, so this will save us a few more bytes and make things a little faster.

Reply:
I was thinking about this a bit more and looking up driver based on Field is really icky, we should really just make `driver` an explicit parameter to this function and make its callers worry about looking this up and passing it in.

`field-should-have-field-values?` is actually called for the batch hydration method for `normal_values` (see `metabase.models.field/with-normal-values`) so you're potentially introducing a not just an `N+1` but a `2N + 1` here with these extra app DB calls (fetch 10 fields, and it instead of doing 1 app DB call it's now doing 21 app DB calls = each Field adds an extra 2 DB calls to get the driver associated with each Field).

This of course is a little annoying to do since we call this function in 7 different places, but I think the only other choice is to have sync record this information which would mean you need to add a new place to store it etc

Reply:
Thanks for thorough explanation. While writing that I had a feeling this may not be the best idea.

The alternative that came to my mind would be to introduce dynamic var and bind that during sync, but that's also something that would need more thought to get that right as there may be some code calling the function that has no knowledge of database.

If we add some initialization to sync as we have in qp (`qp.setup/with-qp-setup`), which I believe may be necessary for completion of https://github.com/metabase/metabase/issues/41216 (not yet sure), it may provide remedy in situations like this one.

However problem is resolved now using the `:type/field-values-unsupported`.

Lesson learned here is to be cautios when adding t2 calls but even more when doing so in sync code.